### PR TITLE
fix gcc-12 -O3 compiler warnings

### LINKED ISSUE #607: fails to build with GCC-12 -O3 (MSYS2 + MinGW)
```cc
[2/15] Building C object src/CMakeFiles/woof.dir/z_zone.c.obj
FAILED: src/CMakeFiles/woof.dir/z_zone.c.obj
D:\msys64\mingw32\bin\cc.exe -DHAVE_FLUIDSYNTH -DMINIZ_NO_ARCHIVE_APIS -DMINIZ_NO_STDIO -DMINIZ_NO_TIME -DRANGECHECK -ID:/msys64/home/FGreffrath/woof/build/src/.. -ID:/msys64/home/FGreffrath/woof/textscreen/. -ID:/msys64/home/FGreffrath/woof/opl/. -isystem D:/msys64/mingw32/include/SDL2 -O3 -DNDEBUG -Wdeclaration-after-statement -Werror=implicit-function-declaration -Werror=incompatible-pointer-types -Werror=int-conversion -Wformat=2 -Wnull-dereference -Wredundant-decls -Wrestrict -Wall -Werror -Dmain=SDL_main -MD -MT src/CMakeFiles/woof.dir/z_zone.c.obj -MF src\CMakeFiles\woof.dir\z_zone.c.obj.d -o src/CMakeFiles/woof.dir/z_zone.c.obj -c D:/msys64/home/FGreffrath/woof/src/z_zone.c
D:/msys64/home/FGreffrath/woof/src/z_zone.c: In function 'Z_Malloc':
D:/msys64/home/FGreffrath/woof/src/z_zone.c:335:32: error: pointer 'block' may be used after 'free' [-Werror=use-after-free]
  335 |             (newb->next = block->next)->prev = newb;
      |                           ~~~~~^~~~~~
In function 'Z_Free',
    inlined from 'Z_Malloc' at D:/msys64/home/FGreffrath/woof/src/z_zone.c:316:10:
D:/msys64/home/FGreffrath/woof/src/z_zone.c:466:10: note: call to 'free' here
  466 |          free(block);
      |          ^~~~~~~~~~~
D:/msys64/home/FGreffrath/woof/src/z_zone.c: In function 'Z_Malloc':
D:/msys64/home/FGreffrath/woof/src/z_zone.c:327:15: error: pointer 'block' may be used after 'free' [-Werror=use-after-free]
  327 |       if(block->tag == PU_FREE && block->size >= size)   // First-fit
      |          ~~~~~^~~~~
In function 'Z_Free',
    inlined from 'Z_Malloc' at D:/msys64/home/FGreffrath/woof/src/z_zone.c:316:10:
D:/msys64/home/FGreffrath/woof/src/z_zone.c:466:10: note: call to 'free' here
  466 |          free(block);
      |          ^~~~~~~~~~~
cc1.exe: all warnings being treated as errors
```

### PR DESCRIPTION
**Pull request by fabiangreffrath**
Fixes #607

### CODE REVIEW

File: src/z_zone.c [hunk 1]
```diff
@@ -310,7 +310,7 @@ void *(Z_Malloc)(size_t size, int tag, void **user, const char *file, int line)
    do
    {
       // Free purgable blocks; replacement is roughly FIFO
-      if(block->tag >= PU_PURGELEVEL)
+      if(block->tag >= PU_PURGELEVEL && !block->vm)
       {
```
Comment (fabiangreffrath) [inline src/z_zone.c:314]: Not too sure about this one.

**Comment by fabiangreffrath** [comment]
We should probably never `free()` stdlib-allocated memory, because this won't play nice with zone memory's concept of a double linked list of memory blocks.

**Comment by rfomin** [comment]
> We should probably never `free()` stdlib-allocated memory, because this won't play nice with zone memory's concept of a double linked list of memory blocks.

Agree. In DSDA-Doom `Z_Malloc` looks much cleaner: https://github.com/kraflab/dsda-doom/blob/fed325a98c650d3f2756eb80c56abd98a25cd620/prboom2/src/z_zone.c#L97

**Comment by fabiangreffrath** [comment]
Yes, but dsda doesn't have the stdlib malloc fallback solution.

**Comment by fabiangreffrath** [comment]
I have some middle ground implemented in Crispy where I increase zone memory at OOM.

**Comment by rfomin** [comment]
> I have some middle ground implemented in Crispy where I increase zone memory at OOM.

Sounds good.
Haven't figured out the details yet to be honest. It would be great to simplify this.

**Comment by fabiangreffrath** [comment]
@kraflab It seems you have tidied up the zone memory implementation in DSDA a fair bit. Could you tell me what happens if you run out of zone memory, i.e. if loading a map requires more memory than you have allocated for the zone? Some maps in Sunder come to mind which required me to add counter measures in Crispy, which are still more a dirty hack than anything else. Will DSDA simply error out with a "failed to allocate..." message?

**Comment by fabiangreffrath** [comment]
Another alternative:
```diff
--- a/src/z_zone.c
+++ b/src/z_zone.c
@@ -312,13 +312,14 @@ void *(Z_Malloc)(size_t size, int tag, void **user, const char *file, int line)
       // Free purgable blocks; replacement is roughly FIFO
       if(block->tag >= PU_PURGELEVEL)
       {
+         const unsigned char vm = block->vm;
          start = block->prev;
          Z_Free((char *) block + HEADER_SIZE);
          /* cph - If start->next == block, we did not merge with the previous
           *       If !=, we did, so we continue from start.
           *  Important: we've reset start!
           */
-         if(start->next == block)
+         if(!vm && start->next == block)
             start = start->next;
          else
             block = start;

```

**Comment by kraflab** [comment]
```c
  while (!(block = (malloc)(size + HEADER_SIZE))) {
    if (!blockbytag[PU_CACHE])
      I_Error ("Z_Malloc: Failure trying to allocate %lu bytes", (unsigned long) size);
    Z_FreeTag(PU_CACHE);
  }
```

Note that prboom-plus and by extension dsda-doom don't actually have memory zones. They just use malloc under the hood without any contiguous blocks. The interface is there for some basic convenience: automatically freeing things with the level tag, error handling, etc. In dsda-doom the main difference is I purged out all the code in `#ifdef 0` blocks and simplified things.

**Comment by fabiangreffrath** [comment]
Thank you! Note to self: Adapt to DSDA's 250 lines z_zone.c for the next release!

**Pull request merged**